### DISCUSSION TITLE
How do I train custom labels?

### DISCUSSION THREAD

**Discussion by ayan4m1** [ultralytics/ultralytics#22547] [Q&A] [answered]
I have yolo11n.pt and the coco8.yaml from [here](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/cfg/datasets/coco8.yaml) - I have a `dataset` directory with labeled train/test/val folders and a dataset.yaml with 1 class, "watermark." (nc is 1 and names is a single element array) I am trying to train this using the YOLO CLI and I get a model which, when converted to ONNX and run through Transformers.js, has 84 output features, four for the bbox and 80 for the COCO classes I do NOT want to train on. Is there a solution to my problem?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ayan4m1, thank you for your interest in Ultralytics 🚀! This is an automated response — an Ultralytics engineer will follow up with you shortly.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including:
-Exact CLI command(s) used to train and export (e.g., `yolo train ...`, `yolo export ...`)  
-Your `dataset.yaml` contents (names/classes, paths)  
-A few labeled sample images from train/val sets  
-Training logs and final results summary  
-Model file you started from (e.g., `yolo11n.pt`) and any config overrides  
-Environment details (OS, Python/PyTorch versions, GPU/CUDA, how installed)  
-What you observe after export (tooling, output shapes, any warnings)  
And please verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) ✅

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
You're probably converting the wrong model. Or you trained the model on the wrong dataset. What's the `coco8.yaml` for in your case if you're training on your own dataset? Can you share the command you used for training and export?

**Reply by ayan4m1** [reply]
I've tried two things:
```
yolo train model=yolo11n.pt data=dataset/data.yaml epochs=10 imgsz=640
```

```
yolo train model=coco8.yaml data=dataset/data.yaml epochs=10 imgsz=640
```

I assume from your post that the latter is incorrect. Is the former incorrect as well?

EDIT: Sorry, for export

```
optimum-cli export onnx --model <my_yolo_model_repo> ./test
```

**Reply by Y-T-G** [reply]
The first one is correct. Why aren't you exporting using Ultralytics YOLO CLI?

**Reply by ayan4m1** [reply]
Sorry, need more coffee.

```
yolo export model=best.pt format=onnx nms=true
```

![labels](https://github.com/user-attachments/assets/712c29d2-c3e3-4c9b-83fc-b45eefaf5f5a)

![train_batch1](https://github.com/user-attachments/assets/d13d6231-6319-47c6-be23-c7dd67ea1c37)

This looks correct, yeah?

**Reply by ayan4m1** [reply]
Trying to use this latest export

```
An error occurred during model execution: "Error: Got invalid dimensions for input: images for the following indices
 index: 2 Got: 960 Expected: 640
 Please fix either the inputs/outputs or the model.".
Inputs given to model: {
  images: {
    type: 'float32',
    dims: [ 1, 3, 960, 640 ],
```

Why does it _demand_ the size be exactly 640?

My preprocessor_config.json

```
"size": {
    "shortest_edge": 640
  },
```

Not sure how this is supposed to work, sorry, this is extremely overwhelming for me. All the documentation says "oh yeah this is great just plug and play everything works together" and the reality is anything but that.

Where am I expected to get all of the misc JSON files that are needed by transformers.js but not provided by the YOLO ONNX export?

And sorry, but I've already been filing issues on transformers.js trying to get this to work, and it's "a problem on the YOLO side" as far as they're concerned. I am stuck here.

[HF repo](https://huggingface.co/ayan4m1/Watermark-Detection-YOLO11-ONNX)

**Reply by ayan4m1** [reply]
Figured maybe I just needed to make it square, so I tried...

```js
const imgBuf = await sharp(sourceBytes).resize(640, 640).toArray();
const image = new RawImage(imgBuf, 640, 640, 3);
const inputs = await processor(image);
```

> Error: Tensor's size(1228800) does not match data length(1)

**Reply by Y-T-G** [reply]
It's plug and play if you use Ultralytics. You can export the ONNX model with a single command, and run prediction with the same ONNX model using Ultralytics with a single command.

If you're using some third party library, then it's not going to be plug and play of course. ONNX is just the model. The preprocessing and postprocessing is part of Ultralytics and it's not something that's exported. It needs to be implemented manually outside of Ultralytics.

The input size has to be exactly the same because that's how ONNX models are by default. If you want dynamic shape, then you just add `dynamic=True` to export which is mentioned in the docs:
https://docs.ultralytics.com/modes/export/#arguments

**Reply by Y-T-G** [reply]
I have never used Transformers.js. But if I am taking a guess, the JSON here is more correct than the one in the repo you linked:

https://github.com/droneforge/yolov11-UAV-finetune/blob/3882117e2a1a4b8e858339080df1dd2d39c86d66/preprocessor.json

**Reply by ayan4m1** [reply]
OK, that makes sense, but even after resizing the image to 640x640 I was getting an error. I will try with dynamic=True and that preprocessor config next. Thanks for your help so far.

**Reply by Y-T-G** [reply]
Did you try the preprocessor JSON I sent above?

**Reply by ayan4m1** [reply]
Yes, sorry, I was trying to say in my previous reply that I needed to try it. It seems like I am getting data back now, but trying to calculate the bounding boxes is not working given some example code that Xenova (the maintainer of transformers.js) provided someone else for YOLO, so something is still wrong. Sorry, but I think I give up. Thanks for your help.

**Reply by glenn-jocher** [reply, marked as answer]
Totally understandable—don’t worry, you’re close. Two things to unblock Transformers.js: export your trained weights without NMS (so you get raw logits) and verify the ONNX itself runs end‑to‑end in Ultralytics before wiring up third‑party pre/postprocess.

```bash
# 1) Re‑export your 1‑class model (no NMS)
yolo export model=best.pt format=onnx dynamic=True nms=False

# 2) Sanity‑check the exported ONNX in Ultralytics
yolo predict model=best.onnx source=path/to/image.jpg imgsz=640
```

For a 1‑class detector the raw ONNX output should be [1, N, 5] = 4 box coords + 1 class score; if you still see 84 you exported the wrong file (e.g., COCO weights). ONNX export options and shapes are summarized in the [ONNX export guide](https://docs.ultralytics.com/integrations/onnx/). If it still fails, please paste the full `yolo export/predict` logs and `yolo checks` so we can dig in.

**Reply by ayan4m1** [reply]
Thanks, turning NMS off made my results sensible again.

**Reply by glenn-jocher** [reply]
Nice, that means you’re now getting the raw logits as expected; from here you just need to handle NMS, score thresholds, and coordinate scaling in Transformers.js, and if you ever hit an Ultralytics-side issue again, feel free to share your `yolo` commands and logs so we can review.
